session/schema_amender.go

```diff
@@ -0,0 +1,532 @@
+	},
+	model.StateWriteOnly: {
+		model.StateWriteReorganization: AmendNone,
+		model.StatePublic:              AmendNone,
+	},
+	model.StateWriteReorganization: {
+		model.StatePublic: AmendNone,
+	},
+}
+
+type schemaAndDecoder struct {
+	schema  *expression.Schema
+	decoder *rowcodec.ChunkDecoder
+}
+
+// amendCollector collects all amend operations.
+type amendCollector struct {
+	tblAmendOpMap map[int64][]amendOp
+}
+
+func newAmendCollector() *amendCollector {
+	res := &amendCollector{
+		tblAmendOpMap: make(map[int64][]amendOp),
+	}
+	return res
+}
+
+func findIndexByID(tbl table.Table, ID int64) table.Index {
+	for _, indexInfo := range tbl.Indices() {
+		if indexInfo.Meta().ID == ID {
+			return indexInfo
+		}
+	}
+	return nil
+}
+
+func findColByID(tbl table.Table, colID int64) *table.Column {
+	for _, colInfo := range tbl.Cols() {
+		if colInfo.ID == colID {
+			return colInfo
+		}
+	}
+	return nil
+}
+
+func addIndexNeedRemoveOp(amendOp int) bool {
+	if amendOp == AmendNeedAddDelete || amendOp == AmendNeedAddDeleteAndInsert {
+		return true
+	}
+	return false
+}
+
+func addIndexNeedAddOp(amendOp int) bool {
+	if amendOp == AmendNeedAddDeleteAndInsert || amendOp == AmendNeedAddInsert {
+		return true
+	}
+	return false
+}
+
+func (a *amendCollector) keyHasAmendOp(key []byte) bool {
+	tblID := tablecodec.DecodeTableID(key)
+	ops := a.tblAmendOpMap[tblID]
+	return len(ops) > 0
+}
+
+func (a *amendCollector) collectIndexAmendOps(sctx sessionctx.Context, tblAtStart, tblAtCommit table.Table) ([]amendOp, error) {
+	res := make([]amendOp, 0, 4)
+	// Check index having state change, collect index column info.
+	for _, idxInfoAtCommit := range tblAtCommit.Indices() {
+		idxInfoAtStart := findIndexByID(tblAtStart, idxInfoAtCommit.Meta().ID)
+		// Try to find index state change.
+		var amendOpType int
+		if idxInfoAtStart == nil {
+			amendOpType = ConstOpAddIndex[model.StateNone][idxInfoAtCommit.Meta().State]
+		} else if idxInfoAtCommit.Meta().State > idxInfoAtStart.Meta().State {
+			amendOpType = ConstOpAddIndex[idxInfoAtStart.Meta().State][idxInfoAtCommit.Meta().State]
+		}
+		if amendOpType != AmendNone {
+			// TODO unique index amend is not supported by now.
+			if idxInfoAtCommit.Meta().Unique {
+				return nil, errors.Trace(errors.Errorf("amend unique index=%v for table=%v is not supported now",
+					idxInfoAtCommit.Meta().Name, tblAtCommit.Meta().Name))
+			}
+			opInfo := &amendOperationAddIndexInfo{}
+			opInfo.AmendOpType = amendOpType
+			opInfo.tblInfoAtStart = tblAtStart
+			opInfo.tblInfoAtCommit = tblAtCommit
+			opInfo.indexInfoAtStart = idxInfoAtStart
+			opInfo.indexInfoAtCommit = idxInfoAtCommit
+			for _, idxCol := range idxInfoAtCommit.Meta().Columns {
+				colID := tblAtCommit.Meta().Columns[idxCol.Offset].ID
+				oldColInfo := findColByID(tblAtStart, colID)
+				// TODO: now index column MUST be found in old table columns, generated column is not supported.
+				if oldColInfo == nil || oldColInfo.IsGenerated() {
```

Comment:
Do we need to consider hidden columns?

Reply:
I'm not quite familiar with the `hidden` column, maybe we should reject this type here?

Reply:
OK~